chalk-engine/src/logic.rs

```diff
@@ -157,38 +195,15 @@ impl<C: Context> Forest<C> {
         );
         info!("table goal = {:#?}", self.tables[table].table_goal);
 
-        // First, check if this table has floundered.
-        if self.tables[table].is_floundered() {
-            return Err(RecursiveSearchFail::Floundered);
-        }
-
-        // Next, check for a tabled answer.
-        if self.tables[table].answer(answer).is_some() {
-            info!("answer cached = {:?}", self.tables[table].answer(answer));
-            return Ok(EnsureSuccess::AnswerAvailable);
-        }
-
-        // If no tabled answer is present, we ought to be requesting
-        // the next available index.
-        assert_eq!(self.tables[table].next_answer_index(), answer);
-
-        // Next, check if the table is already active. If so, then we
-        // have a recursive attempt.
-        if let Some(depth) = self.stack.is_active(table) {
-            info!("ensure_answer: cycle detected at depth {:?}", depth);
-
-            if self.top_of_stack_is_coinductive_from(depth) {
-                return Ok(EnsureSuccess::Coinductive);
-            }
-
-            return Err(RecursiveSearchFail::PositiveCycle(Minimums {
-                positive: self.stack[depth].dfn,
-                negative: DepthFirstNumber::MAX,
-            }));
+        match self.ensure_precheck(table, answer) {
+            Some(res) => return res,
+            None => {}
         }
 
         let dfn = self.next_dfn();
-        let depth = self.stack.push(table, dfn);
+        let work = self.increment_work();
```

Comment:
Seems surprising to me that we implement `work` here -- I feel like that will cause us to do more work in the case of cycles than is warranted. In fact, "work" feels like the wrong thing to be tracking -- aren't we basically just tracking overall answers produced?

Reply:
I think you're correct in that this is the same as overall answers produced. But I need to look more into this. In *most* cases (at least from what I was seeing). `work` tracks `dfn` 1:1. So I need to think about what counter-examples exists and if that matters.